okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java

```diff
@@ -344,6 +347,73 @@ static Status toGrpcStatus(ErrorCode code) {
     return ERROR_CODE_TO_STATUS.get(code);
   }
 
+  private Socket upgradeToTls(Socket socket, SSLSocketFactory sslSocketFactory) throws IOException {
+    SSLSocket sslSocket = (SSLSocket) sslSocketFactory.createSocket(
+        socket, authorityHost, address.getPort(), true /* auto close */);
+    final String[] supportedTlsProtocols = sslSocket.getSupportedProtocols();
+    Set<String> supportedTlsProtocolsSet = new HashSet<String>(supportedTlsProtocols.length);
+    for (int i = 0; i < supportedTlsProtocols.length; ++i) {
+      supportedTlsProtocolsSet.add(supportedTlsProtocols[i]);
+    }
+    List<String> enableTlsProtocols = new ArrayList<String>();
+    addIfSupported(supportedTlsProtocolsSet, enableTlsProtocols,
+        "TLSv1.2", "TLSv1.1", "TLSv1");
+    sslSocket.setEnabledProtocols(
+        enableTlsProtocols.toArray(new String[enableTlsProtocols.size()]));
+
+    final String[] supportedCiphers = sslSocket.getSupportedCipherSuites();
+    Set<String> supportedCiphersSet = new HashSet<String>(supportedCiphers.length);
+    for (int i = 0; i < supportedCiphers.length; ++i) {
+      supportedCiphersSet.add(supportedCiphers[i]);
+    }
+    List<String> enableCiphers = new ArrayList<String>();
+    addIfSupported(
+        supportedCiphersSet, enableCiphers,
```

Comment:
The main takeaway is to "use Java 8," or "Google Play Services Dynamic Security Provider" if on Android.

But we also can allow users to purposefully change the ciphers they are using (for better, or for worse). And thus, we should allow SPEC to be overridden, and even make it public so that users can change only what they need to change.